tests/benchmarks/bench_ceed.cpp

```diff
@@ -0,0 +1,219 @@
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "bench.hpp"
+
+#ifdef MFEM_USE_BENCHMARK
+
+/*
+ This benchmark contains the implementation of the CEED's bake-off problems:
+ high-order kernels/benchmarks designed to test and compare the performance
+ of high-order codes.
+
+ See: ceed.exascaleproject.org/bps and github.com/CEED/benchmarks
+*/
+
+struct BakeOff
+{
+   const int N, p, q, dim = 3;
+   Mesh mesh;
+   H1_FECollection fec;
+   FiniteElementSpace fes;
+   const Geometry::Type geom_type;
+   IntegrationRules IntRulesGLL;
+   const IntegrationRule *irGLL;
+   const IntegrationRule *ir;
+   ConstantCoefficient one;
+   const int dofs;
+   GridFunction x,y;
+   BilinearForm a;
+   double mdofs;
+
+   BakeOff(int p, int vdim, bool GLL):
+      N(Device::IsEnabled()?32:8),
+      p(p),
+      q(2*p + (GLL?-1:3)),
+      mesh(Mesh::MakeCartesian3D(N,N,N,Element::HEXAHEDRON)),
+      fec(p, dim, BasisType::GaussLobatto),
+      fes(&mesh, &fec, vdim),
+      geom_type(fes.GetFE(0)->GetGeomType()),
+      IntRulesGLL(0, Quadrature1D::GaussLobatto),
+      irGLL(&IntRulesGLL.Get(geom_type, q)),
+      ir(&IntRules.Get(geom_type, q)),
+      one(1.0),
+      dofs(fes.GetTrueVSize()),
+      x(&fes),
+      y(&fes),
+      a(&fes),
+      mdofs(0.0) {}
+
+   virtual void benchmark() = 0;
+
+   double SumMdofs() const { return mdofs; }
+
+   double MDofs() const { return 1e-6 * dofs; }
+};
+
+/// Bake-off Problems (BPs)
+template<typename BFI, int VDIM = 1, bool GLL = false>
+struct Problem: public BakeOff
+{
+   const double rtol = 1e-12;
+   const int max_it = 32;
+   const int print_lvl = -1;
+
+   Array<int> ess_tdof_list;
+   Array<int> ess_bdr;
+   LinearForm b;
+   OperatorPtr A;
+   Vector B, X;
+   CGSolver cg;
+
+   Problem(int order):
+      BakeOff(order,VDIM,GLL),
+      ess_bdr(mesh.bdr_attributes.Max()),
+      b(&fes)
+   {
+      ess_bdr = 1;
+      fes.GetEssentialTrueDofs(ess_bdr,ess_tdof_list);
+      b.AddDomainIntegrator(new DomainLFIntegrator(one));
+      b.Assemble();
+
+      a.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      a.AddDomainIntegrator(new BFI(one, GLL?irGLL:ir));
+      a.Assemble();
+      a.Mult(x, y);
+
+      a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
+
+      cg.SetRelTol(rtol);
+      cg.SetOperator(*A);
+      cg.SetMaxIter(max_it);
+      cg.SetPrintLevel(print_lvl);
+      cg.iterative_mode = false;
+      MFEM_DEVICE_SYNC;
+   }
+
+   void benchmark() override
+   {
+      cg.Mult(B,X);
+      MFEM_DEVICE_SYNC;
+      mdofs += MDofs() * cg.GetNumIterations();
+   }
+};
+
+/// Bake-off Problems (BPs)
+#define BakeOff_Problem(i,Kernel,VDIM,p_eq_q)\
+static void BP##i(bm::State &state){\
+   Problem<Kernel##Integrator,VDIM,p_eq_q> ker(state.range(0));\
```

Comment:
Where does `##Integrator` comes from?

Reply:
It's a [concatenation](https://gcc.gnu.org/onlinedocs/cpp/Concatenation.html) with the `Kernel` name.

Reply:
I see, so "Integrator" here wont be substituted and ## is just a separator to have "Kernel" substituted and be able to get something like "DiffusionIntegrator" in the end, right?